dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/events/event/JdbcEventStore.java

```diff
@@ -474,27 +479,131 @@ public JdbcEventStore( StatementBuilder statementBuilder, @Qualifier( "readOnlyJ
         return eventRows;
     }
 
-    private Map<String, String> getDataElementsUidToCode( Set<EventDataValue> eventDataValues, IdScheme idScheme )
+    private Map<String, String> getDataElementsUidToIdentifier( Set<EventDataValue> eventDataValues, IdScheme idScheme )
+    {
+        if ( idScheme.isAttribute() )
+        {
+            return getDataElementsUidToAttribute( eventDataValues );
+        }
+        else
+        {
+            return getDataElementsUidToCode( eventDataValues );
+        }
+    }
+
+    private Map<String, String> getDataElementsUidToCode( Set<EventDataValue> eventDataValues )
     {
         //Get mapping only when needed
-        if ( idScheme != IdScheme.ID && idScheme != IdScheme.UID )
+        Map<String, String> dataElementsUidToCode = new HashMap<>();
+
+        List<String> dataElementUids = eventDataValues.stream()
```

Comment:
Use a Set instead of List here. This way we get the unique data elements (typically a few tens) instead of duplicate data elements for each event (typically in the thousands).

Reply:
As mentioned in Slack, if you check how the `Set<EventDataValue> eventDataValues` is obtained, you will see that it is only dataValues for 1 event extracted from JSONB. In that JSONB we use DataElement UID as a key, so this already takes care of uniqueness. Therefore, changing to Set will not do much difference really. Maybe even make things a bit worse as it needs to check for uniqueness within the Set?

Reply:
Good point.